File: .generate_database/dump.py
#https://stackoverflow.com/questions/39838489/converting-lua-table-to-a-python-dictionary
#https://github.com/SirAnthony/slpp
#pip install git+https://github.com/SirAnthony/slpp
import os
import concurrent.futures
import math
from helpers import find_addon_name, read_expansion_data, get_project_dir_path, get_data_dir_path

addon_dir = find_addon_name()
print(f"Addon dir: {addon_dir}")
print(f"Project dir: {get_project_dir_path()}")

entity_types = ["Quest", "Item", "Object", "Npc"]

expansions = ["Era", "Tbc", "Wotlk"]
range_size = 50

# Both are in number of characters
max_file_size = 45000
max_p_size = 4000

# @dataclass
# class Quest:
#   asdf: int = 0


# The main logic function, it uses the decoded data to create HTML files.
def process_expansion(expansion_name: str, entity_type: str, expansion_data: dict[int, list[str]]):
    from slpp import slpp as lua
    entity_type_lower = entity_type.lower()
    entity_type_plural = entity_type_lower + "s"
    entity_type_capitalized = entity_type.capitalize()
    # A variable to keep track of how many entries have been written
    entries_written = 0
    # Variables to keep track of the lowest and highest id
    lowest_id, highest_id = float('inf'), -float('inf')

    print(f"Processing {expansion_name} python {entity_type_lower} data")


    print(f"Dumping {entity_type_plural} for {expansion_name}")
    path = get_data_dir_path(entity_type, expansion_name)
    data_dir_path= f"Interface\\AddOns\\{addon_dir}\\Database\\{entity_type_capitalized}\\{expansion_name.capitalize()}"

    if not os.path.exists(path + '\\_data'):
        print(f"Creating {entity_type_lower} _data folder")
        os.makedirs(path + '\\_data')
    else:
        #Delete html files in _data folder
        print(f"Deleting old {entity_type_lower} html _data files")
        for file in os.listdir(path + '\\_data'):
            if file.endswith(".html"):
                os.remove(path + '\\_data\\' + file)

    lua.newline = ""
    lua.tab = ""

    # The .xml file that imports all the data files through the toc.
    embed_file_strings = []
    embed_file_strings.append(f'<SimpleHTML name="{entity_type_capitalized}DataFiles" file="{data_dir_path}\\{entity_type_capitalized}DataTemplates.html" virtual="true" font="GameFontNormal"/>\n')


    output_data = ""
    lookup_data_start = "<!-- Index to Id table -->\n" + "<p>"
    # The string that contains the index -> EntityId convertion data printed at the top of each file.
    lookup_data = lookup_data_start
    # Contains all the filenames for the files that are generated, used to create the frames later
    filename_data = ""
    # The indexes of the data that we have written to the file e.g 1 = name, 2 startedBy etc
    writtenDataIndexes = []

    # Keep track of the entry index so we can check if we have reached the end of the dict
    entryIndex = 0
    for dataId in expansion_data:
        entryIndex += 1

        lowest_id = min(lowest_id, int(dataId))
        highest_id = max(highest_id, int(dataId))

        entity_data = expansion_data[dataId]

        output_data_local = ""
        entityDataIndex = 1
        for line in entity_data:
            encoded_line = lua.encode(line)
            encoded_line = encoded_line.replace("\\n", "<br>")
            encoded_line = encoded_line.strip('"')
            #Trim the last comma if it exists
            if encoded_line.endswith(","):
                encoded_line = encoded_line[:-1]
            # Clean up the data
            for i in range(10):
              encoded_line = encoded_line.replace(",}", "}")
              encoded_line = encoded_line.replace("nil}", "}")
            encoded_line = encoded_line.replace(" = ", "=")

            # No reason to print empty lines or nil values
            if encoded_line != "nil" and len(encoded_line) > 0:
                # If the line is too long, we split it into multiple lines
                # We split it at the after max_p_size characters
                if len(encoded_line) < max_p_size:
                    output_data_local += "<p>"
                    output_data_local += encoded_line
                    writtenDataIndexes.append(entityDataIndex)
                    output_data_local += "</p>"
                    output_data_local += "\n"
                else:
                    # If the line is too long, we split it into multiple lines
                    output_data_local += f"<!-- Segment start: {entityDataIndex} -->\n"
                    # Calculate how many segments we need
                    segments = len(encoded_line) / max_p_size
                    segments = math.ceil(segments)

                    # Write each segment as a new tag, ending with e
                    for i in range(segments):
                        if i != segments - 1:
                            writtenDataIndexes.append(f"{entityDataIndex}-{i+1}") # The segment means multiple lines, add one because lua indexes start at 1
                        else:
                            # e as in end
                            writtenDataIndexes.append(f"{entityDataIndex}-{'e'}") # The segment ends at the last line, so we use e to indicate that
                        output_data_local += "<p>"
                        output_data_local += encoded_line[i*max_p_size:min(len(encoded_line), (i+1)*max_p_size)]
                        output_data_local += "</p>\n"
                    output_data_local += "<!-- Segment end: {} -->\n".format(entityDataIndex)
            entityDataIndex += 1

        # Skip the if the entitry is empty
        if len(writtenDataIndexes) > 0:
          lookup_data += f"{dataId},"
          output_data += f"<!-- {dataId} -->\n"
          output_data += "<p>" + ",".join(str(x) for x in writtenDataIndexes) + "</p>\n"
          writtenDataIndexes.clear()
          output_data += output_data_local

          entries_written += 1

        # If we have written the maximum amount of entries, or if we have reached the dict
        if entries_written == range_size or entryIndex == len(expansion_data):# or len(lookup_data) + len(output_data) > max_file_size:
            lookup_data = lookup_data[:-1] + "</p>"
            output_file_name = f"{path}\\_data\\{lowest_id}-{highest_id}.html"
            # if len(lookup_data) + len(output_data) > max_file_size:
            #     print("Warning: File size exceeded for {}-{}.html".format(lowest_id, highest_id))
            #     print(output_file_name)

            output_file = open(output_file_name, 'w', encoding="utf-8")
            output_file.write(f"<html><body>\n{lookup_data}\n{output_data}</body></html>")
            output_file.close()

            embed_file_string = '<SimpleHTML'
            embed_file_string += f' name="{entity_type_capitalized}Data{lowest_id}-{highest_id}"'
            embed_file_string += f' file="{data_dir_path}\\_data\\{lowest_id}-{highest_id}.html"'
            embed_file_string += ' virtual="true"'
            embed_file_string += ' font="GameFontNormal"'
            embed_file_string += '/>\n'
            embed_file_strings.append(embed_file_string)


            filename_data += f"{entity_type_capitalized}Data{lowest_id}-{highest_id},"

            # Reset variables for the next file
            entries_written = 0
            lowest_id, highest_id = float('inf'), -float('inf')
            output_data = ""
            lookup_data = lookup_data_start

        if entryIndex == len(expansion_data) or entryIndex % 1000 == 0:
          print(f"Processed {entryIndex}/{len(expansion_data)} {expansion} {entity_type_plural}")

    # Write out all EntityIds to a file
    with open(f"{path}\\{entity_type_capitalized}DataIds.html", 'w', encoding="utf-8") as entity_id_lookup_file:
      # Write all Ids to a file
      printData = []
      temp = ""
      for Id in expansion_data:
        if len(temp) + len(str(Id)) < max_p_size:
          temp += str(Id) + ","
        else:
          printData.append(temp)
          temp = str(Id) + ","
      printData.append(temp)
      # Remove the last comma in last line
      printData[-1] = printData[-1][:-1]
      entity_id_lookup_file.write(f"<!-- This contains all the ids for {entity_type_plural} -->\n")
      entity_id_lookup_file.write("<html><body>\n")
      for data in printData:
        entity_id_lookup_file.write("<p>" + data + "</p>\n")
      entity_id_lookup_file.write("</body></html>")

    # Add Id lookup file at the start of the embed file
    embed_file_strings.insert(0, f'<SimpleHTML name="{entity_type_capitalized}DataIds" file="{data_dir_path}\\{entity_type_capitalized}DataIds.html" virtual="true" font="GameFontNormal"/>\n')

    with open(f"{path}\\{entity_type_capitalized}DataFiles.xml", 'w', encoding="utf-8") as embed_file:
      embed_file.write('<Ui xsi:schemaLocation="http://www.blizzard.com/wow/ui/ ..\\FrameXML\\UI.xsd">\n')
      for embed_file_string in embed_file_strings:
        embed_file.write(embed_file_string)
      embed_file.write('</Ui>')

    with open(f"{path}\\{entity_type_capitalized}DataTemplates.html", 'w', encoding="utf-8") as filename_file:
      filename_file.write("<!-- This contains all the ranges for the files that are generated -->\n")
      filename_file.write("<html><body>\n")
      #Trim the last comma if it exists
      if filename_data.endswith(","):
          filename_data = filename_data[:-1]
      segments = len(filename_data) / max_p_size
      segments = math.ceil(segments)

      # Write each segment as a new tag, ending with e
      for i in range(segments):
          filename_file.write("<p>")
          filename_file.write(filename_data[i*max_p_size:min(len(filename_data), (i+1)*max_p_size)])
          filename_file.write("</p>\n")
      filename_file.write("</body></html>")

    print(f"Finished dumping {entity_type}s for {expansion_name}")

# Load the data for all expansions
# The decoding with lua is not thread safe, so we do it here
all_expansion_data = {}
for expansion in expansions:
    all_expansion_data[expansion] = {}
    for entity_type in entity_types:
        from slpp import slpp as lua
        print(f"Reading {expansion} lua {entity_type.lower()} data")
        raw_expansion_data = read_expansion_data(expansion, entity_type)
        print(f"Decoding {expansion} lua {entity_type.lower()} data to python {entity_type.lower()}")
        all_expansion_data[expansion][entity_type] = lua.decode(raw_expansion_data)

# Process all expansions in separate threads
with concurrent.futures.ThreadPoolExecutor() as executor:
    futures = [executor.submit(process_expansion, expansion, entity_type, all_expansion_data[expansion][entity_type]) for expansion in expansions for entity_type in entity_types]

    # Wait for all threads to complete
    for future in concurrent.futures.as_completed(futures):
        try:
            future.result()
        except Exception as e:
            print(f"An error occurred in a thread: {e}")

File: .database_generator/dump.lua
-- Allow accessing private fields
---@diagnostic disable: invisible
require("cli.dump")
local helpers = require(".db_helpers")

require("cli.Addon_Meta")
require("cli.CLI_Helpers")

assert(Is_CLI, "This function should only be called from the CLI environment")

local f = string.format

-- Color helper function for terminal output formatting
local c = helpers.colorizeText

---Sanitizes translation strings by replacing special characters with safe HTML entities.
---This is the first phase of string processing - characters are replaced with safe HTML entities
---equivalents to prevent issues during string manipulation and chunking operations.
---@param str string The input string to sanitize
---@return string The sanitized string with special characters replaced by UTF-8 equivalents
local function escapeString(str)
  -- str = str:gsub('&', "and") -- Escape ampersands
  -- str = str:gsub("&", "＆")
  -- str = str:gsub("<", "❮") -- Escape less-than signs
  -- str = str:gsub('>', "❯") -- Escape greater-than signs
  str = str:gsub('&', "＆") -- Escape ampersands
  str = str:gsub('<', "＜") -- Escape less-than signs
  str = str:gsub('>', "＞") -- Escape greater-than signs

  -- We use this "cool" utf-8 dagger because the segmented split can split multi-chars ... 🗡
  str = str:gsub('|n', "🗡")
  str = str:gsub("\\n", "🗡")
  str = str:gsub("\n", "🗡")
  str = str:gsub('\\\\"', '"') -- Unescape double quotes if needed
  return str
end

---Replaces HTML entities with proper HTML entities for final output.
---This is the second phase of string processing - converts the safe HTML entities
---back to proper HTML entities for valid HTML/XML output.
---@param str string The input string with UTF-8 safe characters
---@return string The string with proper HTML entities
local function replaceWithCorrect(str)
  str = str:gsub("🗡", "<br/>") -- Replace 🗡 with <br /> : Use <br /> for XHTML validity
  str = str:gsub('＆', "&amp;") -- Escape ampersands
  str = str:gsub('＜', "&lt;") -- Escape less-than signs
  str = str:gsub('＞', "&gt;") -- Escape greater-than signs
  return str
end

-- ! -------------------------------------------
-- !
-- !  The reason for all these utf-8 things is that chinese characters are multiple "chars" in a lua string
-- !  So if we just do the normal len for strings it would split a chinese character in the middle -
-- !  creating errors ingame.
-- !
-- ! -------------------------------------------
-- UTF-8 char boundary pattern for Lua 5.1 (no \x escapes)
local _CHARPAT = "[%z\1-\127\194-\244][\128-\191]*"

---UTF-8 safe substring function that respects character boundaries.
---Unlike Lua's built-in string.sub which works on bytes, this function works on
---UTF-8 characters to prevent splitting multi-byte characters (like Chinese text).
---@param s string UTF-8 encoded string
---@param i number Start index in characters (1 = first char; negative = from end)
---@param j number? End index in characters (inclusive; positive or negative). If nil, defaults to -1 (last character)
---@return string The UTF-8-safe substring
local function utf8_sub(s, i, j)
  -- 1) collect byte-offsets of each UTF-8 character
  local offsets = {}
  for pos in s:gmatch("()" .. _CHARPAT) do
    offsets[#offsets + 1] = pos
  end
  local n = #offsets
  if n == 0 then return "" end

  -- 2) handle defaults & negative indices
  if not j then j = -1 end
  if i < 0 then i = n + 1 + i end
  if j < 0 then j = n + 1 + j end

  -- 3) clamp to [1..n]
  if i < 1 then i = 1 end
  if j > n then j = n end
  if i > j then return "" end

  -- 4) byte positions for slicing
  local start_byte = offsets[i]
  local end_byte = offsets[j + 1] and (offsets[j + 1] - 1) or #s

  return s:sub(start_byte, end_byte)
end

---Returns the number of UTF-8 characters in a string.
---Unlike Lua's built-in string.len which counts bytes, this function counts
---actual UTF-8 characters for proper length calculation.
---@param s string UTF-8 encoded string
---@return number Count of UTF-8 codepoints/characters
local function utf8_len(s)
  local count = 0
  for _ in s:gmatch(_CHARPAT) do
    count = count + 1
  end
  return count
end

---Generates HTML and XML files for a given entity type and WoW expansion.
---This is the main function that processes database data and converts it into the HTML format
---consumed by the QuestieDB addon. It handles chunking large datasets, UTF-8 safe string
---processing, and generates all necessary auxiliary files for the addon to load the data.
--- @see Corrections.ItemMeta
--- @see Corrections.NpcMeta
--- @see Corrections.ObjectMeta
--- @see Corrections.QuestMeta
---@param dataTbl table<number, table> The final, corrected data table (e.g., itemOverride)
---@param meta ObjectMeta|ItemMeta|QuestMeta|NpcMeta|L10nMeta The metadata table containing keys and dump functions
---@param entityType string The type name ("Item", "Quest", "Npc", "Object", "L10n")
---@param expansionName string The expansion name ("Era", "Tbc", "Wotlk", "Cata")
---@param idsPerFile number? The number of IDs per file chunk (default: 50)
---@param ptagPerFile number? The number of <p> tags per file chunk (default: 65000)
---@param debug boolean? Whether to include debug comments in generated HTML files
function GenerateHtmlForEntityType(dataTbl, meta, entityType, expansionName, idsPerFile, ptagPerFile, debug)
  print(c(f("Generating HTML for %s (%s)...", entityType, expansionName), "green"))

  -- Statistics tracking for file generation analysis
  local p_tags_per_file_avg = 0
  local p_tags_highest = 0
  local p_tags_lowest = math.huge

  -- 1. Define Constants & Paths - Configuration for file generation
  local range_size = idsPerFile or 50          -- Number of entity IDs per HTML file
  local p_tags_per_file = ptagPerFile or 65000 -- Maximum <p> tags per file to prevent memory issues
  local max_file_size = 45000                  -- Less critical now, but good for reference
  local max_p_size = 4000                      -- Maximum characters per <p> tag (forces chunking)
  local entityTypeLower = entityType:lower()
  local entityTypeCapitalized = entityType:gsub("^%l", string.upper)
  local addon_dir = helpers.find_addon_name()
  local outputBasePath = helpers.get_data_dir_path(entityType, expansionName)
  local outputDataPath = outputBasePath .. "/_data"
  -- When doing paths in xml the full path all the way from Interface needs to be used.
  local dataDirPathInAddon = f("Interface\\AddOns\\%s\\Database\\%s\\%s", addon_dir, entityTypeCapitalized, helpers.capitalize(expansionName))

  local fileIndex = 1

  print(f("Addon Directory:       %s", addon_dir))
  print(f("Output Base Path:      %s", outputBasePath))
  print(f("Output Data Path:      %s", outputDataPath))
  print(f("Data Dir Path (Addon): %s", dataDirPathInAddon))

  -- Ensure output directories exist and are prepared
  helpers.ensureDirExists(outputBasePath)
  helpers.ensureDirExists(outputDataPath)

  -- Add .gitkeep file if needed
  helpers.ensureGitKeepFile(outputDataPath)

  -- Clean up old HTML files from previous runs
  local removed_count = helpers.clearHtmlFiles(outputDataPath)
  print(f("Removed %d HTML files from %s", removed_count, outputDataPath))

  -- 2. Initialize State Variables for processing
  local all_ids = {}            -- Track all valid entity IDs for the IDs file
  local sorted_keys = {}        -- Sort entity IDs for consistent file generation
  for k in pairs(dataTbl) do table.insert(sorted_keys, k) end
  table.sort(sorted_keys)       -- Sort IDs numerically

  local all_filenames = {}      -- Track generated filenames for the Templates file
  local embed_file_strings = {} -- Track XML embed strings for the XML file
  table.insert(embed_file_strings,
               f('<SimpleHTML name="%sDataFiles" file="%s\\%sDataTemplates.html" virtual="true" font="GameFontNormal"/>\n', entityTypeCapitalized,
                 dataDirPathInAddon,
                 entityTypeCapitalized))

  -- Chunk processing state variables
  local entries_written           = 0
  local p_tags_written            = 0
  local lowest_id, highest_id     = math.huge, -math.huge
  local current_chunk_output_data = {} -- Use a table for efficient string building
  local current_chunk_lookup_data = {} -- Use a table

  -- Get metadata for processing entity data
  local reversedKeys              = {}
  local nrDataKeys                = 0
  local dataKeys                  = meta[entityTypeLower .. "Keys"] -- e.g., meta.itemKeys
  local dumpFunctions             = meta.dumpFuncs
  local combineFunc               = meta.combine                    -- May be nil
  for key, id in pairs(dataKeys) do
    reversedKeys[id] = key
    nrDataKeys = nrDataKeys + 1
  end

  print(c("Writing HTML files...", "green"))

  -- 3. Main Loop: Process each entity ID and generate HTML content
  for i = 1, #sorted_keys do
    local dataId = sorted_keys[i]
    local entity_data_raw = dataTbl[dataId] -- This is the array of values

    local entity_data_processed = entity_data_raw

    local output_data_local = {}  -- For the current ID's <p> tags
    local writtenDataIndexes = {} -- Tracks indices (1, 2, '3-1', '3-e', etc.)

    -- Prepare the data row using dump functions (similar to dumpData)
    local resulttable = {}
    for keyIndex = 1, nrDataKeys do
      local dataName = reversedKeys[keyIndex]                                              -- Get field name from index
      local dataValue = entity_data_processed[dataName] or entity_data_processed[keyIndex] -- Allow lookup by name or index

      if dataValue ~= nil and dataValue ~= "nil" then
        local dumpFunction = dumpFunctions[dataName]
        if dumpFunction then
          resulttable[keyIndex] = dumpFunction(dataValue) -- Get the Lua-like string representation
        else
          error("No dump function for key: " .. tostring(dataName) .. " (Index: " .. keyIndex .. ")")
        end
      else
        resulttable[keyIndex] = "nil"
      end
    end

    -- Apply combine function if specified (modifies the resulttable structure)
    if combineFunc then
      -- Ensure resulttable has the expected structure for combineFunc
      local tempTableForCombine = {}
      for idx = 1, nrDataKeys do tempTableForCombine[idx] = resulttable[idx] or "nil" end
      combineFunc(tempTableForCombine)
      -- Update resulttable from tempTableForCombine (indices might change!)
      -- This needs careful mapping based on combineFunc's behavior.
      -- Assuming combineFunc modifies the table passed to it:
      resulttable = tempTableForCombine
    end

    -- Inner Loop: Process each data field and generate HTML <p> tags
    for entityDataIndex = 1, #resulttable do -- Use the length of the potentially combined table
      local serialized_line = resulttable[entityDataIndex]

      -- Skip nil/empty entries *after* serialization
      if serialized_line ~= "nil" and serialized_line ~= "" then
        -- HTML Formatting and string sanitization
        local formatted_line = tostring(serialized_line)

        -- Remove surrounding quotes added by dump functions
        -- The reason for this is that the dump function adds ' to the start and end of the string
        ---@see DumpFunctions.dump
        if formatted_line:sub(1, 1) == "'" and formatted_line:sub(-1) == "'" then
          formatted_line = formatted_line:sub(2, -2)
        end

        -- Apply first phase of string sanitization (UTF-8 safe characters)
        -- -- formatted_line = formatted_line:gsub('&', "and") -- Escape ampersands
        -- formatted_line = formatted_line:gsub("&", "＆")
        -- formatted_line = formatted_line:gsub("<", "❮") -- Escape less-than signs
        -- formatted_line = formatted_line:gsub('>', "❯") -- Escape greater-than signs
        -- -- formatted_line = formatted_line:gsub('<', "lt;")    -- Escape less-than signs
        -- -- formatted_line = formatted_line:gsub('>', "gt;")    -- Escape greater-than signs
        -- -- We use this cool dagger because the segmented split can split multi-chars ... 🗡
        -- formatted_line = formatted_line:gsub('|n', "<br/>")
        -- formatted_line = formatted_line:gsub("\\n", "<br/>") -- Use <br/> for XHTML validity if needed
        -- formatted_line = formatted_line:gsub("\n", "<br/>")  -- Use <br/> for XHTML validity if needed
        -- formatted_line = formatted_line:gsub('\\\\"', '"')   -- Unescape double quotes if needed
        formatted_line = escapeString(formatted_line)

        -- Remove surrounding quotes ONLY if not a table literal
        if not formatted_line:match("^%s*{") and not formatted_line:match("}%s*$") then
          if formatted_line:match('^"') and formatted_line:match('"$') then
            formatted_line = formatted_line:sub(2, -2)
          end
          -- Unescape single quotes potentially added by DumpFunctions.dump
          formatted_line = formatted_line:gsub("\\'", "'")
        end

        -- Cleanups (less critical if DumpFunctions are precise)
        -- formatted_line = formatted_line:gsub(",}", "}")
        -- formatted_line = formatted_line:gsub("nil}", "}")
        -- formatted_line = formatted_line:gsub(" = ", "=")

        -- Check length and split if necessary to prevent overly large <p> tags
        local length = #formatted_line > max_p_size and utf8_len(formatted_line) or #formatted_line
        if length > max_p_size then
          -- Large content needs to be split into multiple <p> tags
          if debug then
            table.insert(output_data_local, f("  <!-- %s -->\n", meta.NameIndexLookupTable[entityDataIndex]))
          end
          if debug then
            table.insert(output_data_local, f("<!-- Segment start: %s -->\n", entityDataIndex))
          end
          local segments = math.ceil(length / max_p_size)
          for seg = 1, segments do
            local start = (seg - 1) * max_p_size + 1
            local stop = math.min(seg * max_p_size, length)
            local segmentMarker
            if seg < segments then
              segmentMarker = f("%s-%d", entityDataIndex, seg)
            else
              segmentMarker = f("%s-e", entityDataIndex) -- End marker
            end
            p_tags_written = p_tags_written + 1
            table.insert(writtenDataIndexes, segmentMarker)
            table.insert(output_data_local, f("<p>%s</p>\n", replaceWithCorrect(utf8_sub(formatted_line, start, stop))))
          end
          if debug then
            table.insert(output_data_local, f("<!-- Segment end: %s -->\n", entityDataIndex))
          end
        else
          -- Add the single <p> tag for normal-sized content
          table.insert(writtenDataIndexes, tostring(entityDataIndex))
          if (debug) then
            table.insert(output_data_local, f("  <!-- %s -->\n", meta.NameIndexLookupTable[entityDataIndex]))
          end
          p_tags_written = p_tags_written + 1
          table.insert(output_data_local, f("<p>%s</p>\n", replaceWithCorrect(formatted_line)))
        end
      end
    end -- End inner loop (data fields)

    -- If data was written for this ID, add it to the current chunk
    if #writtenDataIndexes > 0 then
      table.insert(all_ids, dataId) -- Track all valid IDs

      -- Update chunk state tracking
      lowest_id = math.min(lowest_id, dataId)
      highest_id = math.max(highest_id, dataId)
      entries_written = entries_written + 1

      -- Calculate p_tags stats
      p_tags_written = p_tags_written + 1
      if p_tags_written > p_tags_highest then
        p_tags_highest = p_tags_written
      end
      if p_tags_written < p_tags_lowest then
        p_tags_lowest = p_tags_written
      end

      -- Add this entity's data to the current chunk
      table.insert(current_chunk_lookup_data, tostring(dataId))
      table.insert(current_chunk_output_data, f("<!-- %d -->\n", dataId))      -- Add ID comment
      table.insert(current_chunk_output_data, "<p>" .. table.concat(writtenDataIndexes, ",") .. "</p>\n")
      table.insert(current_chunk_output_data, table.concat(output_data_local)) -- Add the generated <p> tags
    end

    -- Check if current chunk should be written to file
    -- Triggers: reached max IDs per file, processed all IDs, or hit <p> tag limit
    if entries_written == range_size or i == #sorted_keys or p_tags_written >= p_tags_per_file then
      if entries_written > 0 then -- Don't write empty chunks
        -- local chunk_filename = f("%d-%d.html", lowest_id, highest_id) -- This is the old way of doing it, it wrote the id range.f
        -- Generate chunk filename and metadata
        local chunk_filename = f("%d.html", fileIndex)
        local chunk_filepath = outputDataPath .. "/" .. chunk_filename
        local chunk_frame_name = f("%sData%d-%d", entityTypeCapitalized, lowest_id, highest_id)

        -- print(f("  Writing chunk: %s (%s entries)", chunk_filename, entries_written))

        -- Write the chunk file with HTML structure
        local file = io.open(chunk_filepath, "w")
        if file then
          file:write("<html><body>\n")
          -- Write Index -> ID lookup <p> (first <p> tag contains ID mapping)
          file:write("<!-- Index to Id table -->\n")
          file:write("<p>" .. table.concat(current_chunk_lookup_data, ",") .. "</p>\n")
          -- Write actual entity data
          file:write(table.concat(current_chunk_output_data))
          file:write("</body></html>\n")
          file:close()

          -- Track filename for Templates file and XML embed generation
          table.insert(all_filenames, chunk_frame_name)
          -- Add embed string for XML file
          local embed_str = f('<SimpleHTML name="%s" file="%s\\_data\\%s" virtual="true" font="GameFontNormal"/>\n',
                              chunk_frame_name, dataDirPathInAddon, chunk_filename)
          table.insert(embed_file_strings, embed_str)
        else
          error("Failed to open file for writing: " .. chunk_filepath)
        end

        -- Update statistics
        p_tags_per_file_avg = (p_tags_per_file_avg + p_tags_written) / 2

        -- Reset chunk state for next chunk
        fileIndex = fileIndex + 1
        entries_written = 0
        p_tags_written = 0
        lowest_id, highest_id = math.huge, -math.huge
        current_chunk_output_data = {}
        current_chunk_lookup_data = {}
      end
    end
  end
  -- End main loop (IDs)

  -- Report file generation statistics
  fileIndex = fileIndex - 1 -- Last fileIndex is not written, so we need to subtract 1
  print(f("  Created files: %d.html - %d.html", 1, fileIndex))
  if fileIndex - removed_count > 0 then
    print(c(f("  Added: %d files.", fileIndex - removed_count), "yellow"))
  elseif fileIndex - removed_count < 0 then
    print(c(f("  Removed: %d files.", removed_count - fileIndex), "yellow"))
  end

  -- 4. Generate Auxiliary Files required by the addon

  print(c(f("Writing Meta Files for: %s (%s)", entityType, expansionName), "green"))

  -- Write ID File (<Type>DataIds.html) - Contains all valid entity IDs
  local ids_filepath = outputBasePath .. "/" .. entityTypeCapitalized .. "DataIds.html"
  local ids_file = io.open(ids_filepath, "w")
  if ids_file then
    ids_file:write(f("<!-- This contains all the ids for %ss -->\n", entityType))
    ids_file:write("<html><body>\n")
    local current_p = {}
    local current_len = 0
    for _, id in ipairs(all_ids) do
      local id_str = tostring(id)
      local id_len = #id_str + 1                                                 -- Include comma
      if current_len + id_len > max_p_size and #current_p > 0 then
        ids_file:write("<p>" .. table.concat(current_p, ",") .. "," .. "</p>\n") -- Always end with a , because we split on ,
        current_p = {}
        current_len = 0
      end
      table.insert(current_p, id_str)
      current_len = current_len + id_len
    end
    if #current_p > 0 then                                                     -- Write remaining IDs
      ids_file:write("<p>" .. table.concat(current_p, ",") .. "," .. "</p>\n") -- Always end with a , because we split on ,
    end
    ids_file:write("</body></html>\n")
    ids_file:close()
    print(f("  Written ID file: %s", ids_filepath))
    -- Add embed string for XML file (insert at beginning)
    table.insert(embed_file_strings, 1,
                 f('<SimpleHTML name="%sDataIds" file="%s\\%sDataIds.html" virtual="true" font="GameFontNormal"/>\n', entityTypeCapitalized, dataDirPathInAddon,
                   entityTypeCapitalized))
  else
    error("Failed to open file for writing: " .. ids_filepath)
  end

  -- Write Template File (<Type>DataTemplates.html) - Contains references to all chunk files
  local templates_filepath = outputBasePath .. "/" .. entityTypeCapitalized .. "DataTemplates.html"
  local templates_file = io.open(templates_filepath, "w")
  if templates_file then
    templates_file:write("<!-- This contains all the ranges for the files that are generated -->\n")
    templates_file:write("<html><body>\n")
    local full_template_string = table.concat(all_filenames, ",")
    local current_pos = 1
    -- Split template string into chunks to avoid overly large <p> tags
    while current_pos <= #full_template_string do
      local end_pos = math.min(current_pos + max_p_size - 1, #full_template_string)
      -- Avoid splitting mid-filename if possible (optional refinement)
      templates_file:write("<p>" .. full_template_string:sub(current_pos, end_pos) .. "</p>\n")
      current_pos = end_pos + 1
    end
    templates_file:write("</body></html>\n")
    templates_file:close()
    print(f("  Written Templates file: %s", templates_filepath))
  else
    error("Failed to open file for writing: " .. templates_filepath)
  end

  -- Write XML File (<Type>DataFiles.xml) - Contains SimpleHTML definitions for addon loading
  local xml_filepath = outputBasePath .. "/" .. entityTypeCapitalized .. "DataFiles.xml"
  local xml_file = io.open(xml_filepath, "w")
  if xml_file then
    xml_file:write(
      '<Ui xmlns="http://www.blizzard.com/wow/ui/" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.blizzard.com/wow/ui/ https://raw.githubusercontent.com/Gethe/wow-ui-source/live/Interface/AddOns/Blizzard_SharedXML/UI.xsd">\n')
    for _, embed_str in ipairs(embed_file_strings) do
      xml_file:write(embed_str)
    end
    xml_file:write("</Ui>\n")
    xml_file:close()
    print(f("  Written XML file: %s", xml_filepath))
  else
    error("Failed to open file for writing: " .. xml_filepath)
  end

  -- Report final statistics
  print(f("Average <p> tags per file: Avg: %d , High: %d , Low: %d", p_tags_per_file_avg, p_tags_highest, p_tags_lowest))
  print(c(f("Finished HTML generation for %s (%s).", entityType, expansionName), "green"))
end -- End GenerateHtmlForEntityType
